File: solution/microservices/ad_engine/src/domain/schemas/response/stats.rs
#[derive(Debug, serde::Serialize, utoipa::ToSchema, validator::Validate)]
#[schema(title = "Stats", description = "View stat")]
pub struct StatResponse {
    #[schema(example = 75)]
    pub impressions_count: u32,
    #[schema(example = 25)]
    pub clicks_count: u32,
    #[schema(example = 33.3)]
    pub conversion: f64,
    #[schema(example = 5550.0)]
    pub spent_impressions: f64,
    #[schema(example = 3550.0)]
    pub spent_clicks: f64,
    #[schema(example = 9100.0)]
    pub spent_total: f64,
}

#[derive(Debug, Default, Clone, serde::Serialize, utoipa::ToSchema, validator::Validate)]
#[schema(title = "Stat by day", description = "View stats by day")]
pub struct StatDailyResponse {
    #[schema(example = 75)]
    pub impressions_count: u32,
    #[schema(example = 25)]
    pub clicks_count: u32,
    #[schema(example = 33.3)]
    pub conversion: f64,
    #[schema(example = 5550.0)]
    pub spent_impressions: f64,
    #[schema(example = 3550.0)]
    pub spent_clicks: f64,
    #[schema(example = 9100.0)]
    pub spent_total: f64,
    #[schema(example = 1)]
    pub date: u32,
}

File: solution/microservices/telegram_bot/src/services/ad_engine/schemas/stats.py
from pydantic import BaseModel, Field


class StatDailyResponse(BaseModel):
    impressions_count: int = Field(..., ge=0)
    clicks_count: int = Field(..., ge=0)
    conversion: float
    spent_impressions: float
    spent_clicks: float
    spent_total: float
    date: int = Field(..., ge=0)


class StatResponse(BaseModel):
    impressions_count: int = Field(..., ge=0)
    clicks_count: int = Field(..., ge=0)
    conversion: float
    spent_impressions: float
    spent_clicks: float
    spent_total: float
